In [8]:
# Import necessary packages
from sqlalchemy import create_engine, inspect
import pandas as pd
import re


In [2]:
# Define your connection parameters
server = 'serverHere'
database = 'sandbox'

# Create an engine
engine = create_engine(f'mssql+pyodbc://{server}/{database}?driver=ODBC+Driver+18+for+SQL+Server&TrustServerCertificate=yes')

In [3]:

# Create an inspector
inspector = inspect(engine)

# Get a list of all table names
tables = inspector.get_table_names()


In [4]:
# Display the list of tables
print("Tables in the database:")
for table in tables:
    print(table)

# Dictionary to hold DataFrames and list to hold DataFrame names
dataframes = {}
dataframe_names = []

Tables in the database:
additional_product
additional_service
city_state_zip
customer
customer_to_clean
oil_flavor
product
product_order
product_type


In [5]:
# Fetch the customer table into a DataFrame
customer_to_clean_df = pd.read_sql('SELECT * FROM customer_to_clean', engine)
print("Original DataFrame:")
print(customer_to_clean_df.head())



Original DataFrame:
   customer_id first_name last_name        full_name    phone_number  \
0            1      Cally  Reynolds   Cally Reynolds  (901) 166-8355   
1            2     Sydney  Bartlett  Sydney Bartlett  (982) 231-7357   
2            3     Hunter    Newton    Hunter Newton  (831) 996-1240   
3            4     Brooke   Perkins   Brooke Perkins  (340) 732-9367   
4            5      Nolan    Slater     Nolan Slater  (540) 487-5928   

         street_address  city_state_zip_id  
0     556 Lakewood Park                  8  
1   4829 Badeau Parkway                 14  
2  2 Rockefeller Avenue                 85  
3     87 Brentwood Park                 23  
4        99 Sage Street                 59  


In [6]:
# # Function to clean phone numbers
# def clean_phone_number(phone):
#     # Remove all non-numeric characters
#     cleaned = re.sub(r'\D', '', phone)
#     return cleaned



In [15]:
# # Apply the function to create a new column
# customer_to_clean_df['New_phone_number'] = customer_to_clean_df['phone_number'].apply(clean_phone_number)

# print("\nDataFrame with New Phone Numbers:")
# print(customer_to_clean_df[['phone_number', 'New_phone_number']].head())




DataFrame with New Phone Numbers:
     phone_number New_phone_number
0  (901) 166-8355       9011668355
1  (982) 231-7357       9822317357
2  (831) 996-1240       8319961240
3  (340) 732-9367       3407329367
4  (540) 487-5928       5404875928


In [ ]:
# # Option 1: Update the existing table
# # customer_df.to_sql('customer', engine, if_exists='replace', index=False)
# # print("\nUpdated customer table with new phone numbers.")

# # Option 2: Create a new table
# customer_to_clean_df.to_sql('customer_cleaned', engine, if_exists='replace', index=False)
# print("\nCreated a new table 'customer_cleaned' with cleaned phone numbers.")

Other option

In [10]:
# Function to clean phone numbers by removing (, ), and -
def clean_phone_number(phone):
    # Remove the characters (, ), and -
    cleaned = re.sub(r'[()\-\s]', '', phone)
    return cleaned

In [11]:
# Apply the function to create a new column
customer_to_clean_df['New_phone_number'] = customer_to_clean_df['phone_number'].apply(clean_phone_number)



In [12]:
print("\nDataFrame with New Phone Numbers:")
print(customer_to_clean_df[['phone_number', 'New_phone_number']].head())




DataFrame with New Phone Numbers:
     phone_number New_phone_number
0  (901) 166-8355       9011668355
1  (982) 231-7357       9822317357
2  (831) 996-1240       8319961240
3  (340) 732-9367       3407329367
4  (540) 487-5928       5404875928


In [14]:
# Option 1: Update the existing table
# customer_df.to_sql('customer', engine, if_exists='replace', index=False)
# print("\nUpdated customer table with new phone numbers.")

# Option 2: Create a new table
customer_to_clean_df.to_sql('customer_cleaned', engine, if_exists='replace', index=False)
print("\nCreated a new table 'customer_cleaned' with cleaned phone numbers.")


Created a new table 'customer_cleaned' with cleaned phone numbers.


Data Cleaning Process

Introduction
The goal of this task was to clean the phone_number column in the customer table by removing non-numeric characters such as parentheses ( and ), hyphens -, and any spaces. This cleaned data was then stored in a new column called New_phone_number. Finally, the cleaned data was saved to a new table in the sandbox database on the server.

Methodology
Step-by-Step Process:

Connect to the Database:

Using SQLAlchemy, we connected to the SQL Server database.
Load the Data:

The customer table was loaded into a pandas DataFrame.
Clean the Data:

We defined a function to remove the specified non-numeric characters from the phone_number column.
Applied this function to create a new column, New_phone_number.
Save the Cleaned Data:

The cleaned DataFrame, containing both the original and new phone number columns, was saved to a new table in the database.
Results
After running the notebook:

The phone_number column in the customer table was successfully cleaned.
A new column, New_phone_number, was created in the DataFrame to store the cleaned phone numbers.
The cleaned data was saved to a new table named customer_cleaned in the sandbox database on the server.
Example of Cleaned Data
phone_number	New_phone_number
(901) 166-8355	9011668355
(982) 231-7357	9822317357
(831) 996-1240	8319961240
(340) 732-9367	3407329367
(540) 487-5928	5404875928
Conclusion
By cleaning the phone_number column and creating a new table in the sandbox database, we have:

Improved data quality by standardizing phone numbers.
Ensured that the data is ready for further analysis or integration into other systems.